cirq/linalg/transformations.py

```diff
@@ -304,3 +317,152 @@ def partial_trace(tensor: np.ndarray,
     left_indices = [keep_map[i] if i in keep_set else i for i in range(ndim)]
     right_indices = [ndim + i if i in keep_set else i for i in left_indices]
     return np.einsum(tensor, left_indices + right_indices)
+
+
+def wavefunction_partial_trace(
+        wavefunction: np.ndarray,
+        keep_indices: List[int],
+        *,
+        atol: Union[int, float] = 1e-8) -> Tuple[Tuple[float, np.ndarray]]:
+    """Attempts to factor a wavefunction into two parts and return one of them.
+
+    The input wavefunction must have shape `(2, 2, ..., 2)` or `(2 ** N)` where
+    `wavefunction` is expressed over N qubits. Attempts to factor the
+    wavefunction into a pure state over `keep_indices`. If this fails, a non-
+    unique mixture can be computed by eigendecomposition on the partial trace of
+    the input's density matrix.
+
+    States in the output mixture will retain the same type of shape as the input
+    wavefunction, either `(2 ** k)` or `(2, 2, ..., 2)` where k is the number of
+    qubits kept.
+
+    Args:
+        wavefunction: A wavefunction to express over a qubit subset.
+        keep_indices: Which indices to express the wavefunction on.
+        atol: The tolerance for determining that a factored state is pure.
+
+    Returns:
+        A single-component mixture in which the factored wavefunction has
+        probability '1' if the factored state is pure, or else a mixture of the
+        default eigendecomposition of the mixed state's partial trace.
+
+    Raises:
+        ValueError: if the input wavefunction shape is neither `(2 ** N)` nor
+        `(2, 2, ..., 2)`
+    """
+
+    # Attempt to do efficient state factoring.]
+    state = subwavefunction(wavefunction, keep_indices, default=None, atol=atol)
+    if state is not None:
+        return ((1.0, state),)
+
+    # Fall back to a (non-unique) mixture representation.
+    if wavefunction.shape == (wavefunction.size,):
+        return_shape = -1
+    elif wavefunction.shape == (2,) * int(np.log2(wavefunction.size)):
```

Comment:
`elif all(e == 2 for e in wavefunction.shape):`

Or add 0.5 to the log output, to avoid the assumption that there is no floating point error in the log2 for powers of 2.